### PR TITLE
Add Kotlin Language Support

### PR DESCRIPTION
**Pull request by creativedrewy**
Here is my first attempt at adding Kotlin sytnax highlighting. Is there a way I can test this online in the highlighter preview? I'd like to try out some permutations.

### CODE REVIEW

**Comment by langpavel** [comment]
Hi @creativedrewy 

As I don't know kotlin, I can help you with testing of it :-)

 - `mkdir workspace && cd workspace`
 - clone [monaco-languages](https://github.com/Microsoft/monaco-languages)
 - `cd monaco-languages`
 - `npm run prepublishOnly`
 - `npm link`
 - clone [monaco-editor](https://github.com/Microsoft/monaco-editor)
 - `npm link monaco-languages`
 - `npm run release` *(there are also `website` and `build-website` tasks, not know exactly what is needed)*
 - `npm run simpleserver`
- navigate to http://localhost:8080/monaco-editor/test/ and select in *Mode*, you should see `kotlin` here

**Comment by creativedrewy** [comment]
This was super helpful! I was able to see Kotlin in the web UI and it is initially working. The one thing that I can't seem to figure out is how to get class name highlighting. I've tried adding:

```typescript
[/[A-Z][\w\$]*/, 'type.identifier'],
```

a couple of places in the tokenizer sub-region, but it just won't seem to highlight for me. Is there documentation anywhere about how to use all the options of the tokenzier? I used the javascript file as  a template. On second thought that maybe wasn't the best template :).

Also, once this PR is accepted, should I add a kotlin sample to the monaco editor?

**Comment by langpavel** [comment]
Order matters, so if you have more generic expression before your `type.identifier`, it can eat it. Try add new rule on top

File: src/kotlin/kotlin.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	defaultToken: '',
+	tokenPostfix: '.kt',
+
+	keywords: [
+		'as', 'as?', 'break', 'class', 'continue', 'do', 'else', 'false', 'for', 'fun', 'if',
+		'in', '!in', 'interface', 'is', '!is', 'null', 'object', 'package', 'return', 'super',
+		'this', 'throw', 'true', 'try', 'typealias', 'val', 'var', 'when', 'while', 'by',
+		'catch', 'constructor', 'delegate', 'dynamic', 'field', 'file', 'finally', 'get',
+		'import', 'init', 'param', 'property', 'receiver', 'set', 'setparam', 'where', 'actual',
+		'abstract','annotation', 'companion', 'const', 'crossinline', 'data', 'enum', 'expect',
```
Comment (caltseng) [inline src/kotlin/kotlin.ts]: When is this used in Kotlin? When I tried this, I constantly had my imports highlighted, which didn't look right. 

I'm just not as familiar with this.

File: src/kotlin/kotlin.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+		comment: [
+			[/[^\/*]+/, 'comment'],
+			[/\*\//, 'comment', '@pop'],
+			[/[\/*]/, 'comment']
+		],
+		//Identical copy of comment above, except for the addition of .doc
+		javadoc: [
+			[/[^\/*]+/, 'comment.doc'],
+			// [/\/\*/, 'comment.doc', '@push' ],    // nested comment not allowed :-(
```
Comment (caltseng) [inline src/kotlin/kotlin.ts]: Nit: delete comment

File: src/kotlin/kotlin.test.ts [hunk 3]
```diff
@@ -0,0 +1,647 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+import { testTokenization } from '../test/testRunner';
+
+testTokenization('kotlin', [
```
Comment (caltseng) [inline src/kotlin/kotlin.test.ts]: I'll be adding a couple more tests here that I wrote for the following lines: 

```
inline fun <reified T : Any> foo()',

'val x: X=5'
```

Comment (creativedrewy) [reply src/kotlin/kotlin.test.ts]: 👍

**Comment by creativedrewy** [comment]
This looks awesome! I have confirmed that class highlighting works. Can we get this merged?

**Comment by alexandrudima** [comment]
:+1: Thank you!

**Comment by alexandrudima** [comment]
The steps from @langpavel also get the job done. A slightly simpler version that I use (which doesn't involve npm linking):

 - `mkdir workspace && cd workspace`
 - clone [monaco-languages](https://github.com/Microsoft/monaco-languages), `cd monaco-languages`, `npm install` and `npm run prepublishOnly`
 - clone [monaco-editor](https://github.com/Microsoft/monaco-editor), `cd monaco-editor`, `npm install`, `npm run simpleserver`
 - navigate to `http://localhost:8080/monaco-editor/test/` and **switch to run monaco-languages from source** by clicking on the `src` next to it
 - this will transform the URL to `http://localhost:8080/monaco-editor/test/?monaco-languages=src`

image

You can confirm the correct location is loaded by either seeing your language available in the popup or by looking at the console which echoes the loading paths:
```
{
	"vs/language/typescript": "/monaco-editor/node_modules/monaco-typescript/release/dev",
	"vs/language/css": "/monaco-editor/node_modules/monaco-css/release/dev",
	"vs/language/json": "/monaco-editor/node_modules/monaco-json/release/dev",
	"vs/language/html": "/monaco-editor/node_modules/monaco-html/release/dev",
	"vs/basic-languages": "/monaco-languages/release/dev",
	"vs": "/monaco-editor/node_modules/monaco-editor-core/dev/vs"
}
```

**Pull request merged**